src/tribler-common/tribler_common/utilities.py

```diff
@@ -27,6 +29,46 @@ def uri_to_path(uri):
 
 
 fts_query_re = re.compile(r'\w+', re.UNICODE)
+extract_tags_re = re.compile(r'\[\w+\]')
+
+
+@dataclass
+class Query:
+    original_query: Optional[str]
+    tags: Set[str] = field(default_factory=set)
+    fts_text: str = ''
+
+
+def parse_query(query: Optional[str]) -> Query:
+    """
+    The query structure:
+        query = [tag1][tag2] text
+                 ^           ^
+                tags        fts query
+    """
+    if not query:
+        return Query(original_query=query)
+
+    tags = extract_tags(query)
+    fts_text = extract_plain_fts_query_text(query, tags)
+
+    return Query(original_query=query, tags=tags, fts_text=fts_text)
+
+
+def extract_tags(text: Optional[str]) -> Set[str]:
+    if not text:
+        return set()
+    return {tag[1:-1] for tag in extract_tags_re.findall(text)}
+
+
+def extract_plain_fts_query_text(query: Optional[str], tags: Optional[Set[str]]) -> str:
+    if query is None or tags is None:
```

Comment:
Agree with tags, disagree with query, as it comes from `self.top_search_bar.text()`: https://github.com/drew2a/tribler/blob/465c7be7b093dd9708d96c3ea1b84198208083a7/src/tribler-gui/tribler_gui/tribler_window.py#L1034

Reply:
`Optional` for tags has been removed.

Reply:
> Agree with tags, disagree with query, as it comes from `self.top_search_bar.text()`

**nitpick** But should the function API depend on such external details? If `self.top_search_bar.text()` can actually be None, we can pass it to the function as `parse_query(self.top_search_bar.text() or "")`.

It's like having a function `add_two_ints(a, b)` with `a` argument type defined as `Union[int, str]` just because someone passes `"123"` to the function as a string without proper conversion to `int` first. It is logically clearer to convert the argument to the right type before the function call, not inside it.

Actually, if I read [the documentation](https://doc.qt.io/qt-5/qlineedit.html#text-prop) of QLineEdit correctly, the value of `text()` should always be string (an empty string if the field is empty):

Reply:
Fixed.